### PR TITLE
Fix NuGet search endpoints

### LINKED ISSUE #23191: NuGet v3 Query api take maximum value or 1000
### Feature Description

NuGet UI Manager of VS2022 only get latest few version list from gitea api returns 10 items.

The api not support MAX_RESPONSE_ITEMS and DEFAULT_PAGING_NUM.

see: https://learn.microsoft.com/en-us/nuget/api/search-query-service-resource?source=recommendations
> The skip parameter defaults to 0.
> The take parameter should be an integer greater than zero. The server implementation may impose a maximum value.
>  Note
> nuget.org limits the skip parameter to 3,000 and the take parameter to 1,000.

### Screenshots

_No response_

### LINKED ISSUE #25564: Nuget Package Explorer only loads several nuget packages
### Description

I have uploaded about 2900 Nuget packages to a local installed gitea instance, which worked fine. I can see all packages in the package list of the organisation in the Web-UI an also search for specific Packages.
But if I try to load all Packages e.g. in Visual Studio or in nuget Packages Explorer, i only some several of these packages:
image

In gitea Web-UI it looks like this:
image

Anybody knows this issue?
I have not testet it on try.gitea.io.

### Gitea Version

1.20.0+rc2-12-g102dcfa3a

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

Docker on Debian

### How are you running Gitea?

As docker image

### Database

PostgreSQL

### PR DESCRIPTION
**Pull request by KN4CK3R**
Fixes #25564
Fixes #23191

- Api v2 search endpoint should return only the latest version matching the query
- Api v3 search endpoint should return `take` packages not package versions

### CODE REVIEW

File: models/db/context.go [hunk 1]
```diff
@@ -264,3 +266,29 @@ func inTransaction(ctx context.Context) (*xorm.Session, bool) {
...
+	switch t := GetEngine(ctx).(type) {
+	case *xorm.Engine:
+		e = t
+	case *xorm.Session:
+		e = t.Engine()
+	default:
+		panic(fmt.Sprintf("unexpected engine type: %T", t))
+	}
+
+	switch e.Dialect().URI().DBType {
```
Comment (lunny) [inline models/db/context.go]: We can get this by `setting.Database.Type.IsMySQL()` and etc.

File: models/packages/nuget/search.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+	total, err := e.
+		Where(cond).
+		Count(&packages_model.Package{})
+	if err != nil {
+		return nil, 0, err
+	}
+
+	inner := builder.
+		Dialect(db.BuilderDialect()).
```
Comment (lunny) [inline models/packages/nuget/search.go]: Why the dialect is necessary? Since there is no limitation as condition, I think the dialect could be ignored.

Comment (KN4CK3R) [reply models/packages/nuget/search.go]: It's in line 48.

**Comment by KN4CK3R** [comment]
The new build error seems related to the xorm changes:
```
[E] PANIC: interface conversion: builder.Writer is *statements.QuoteReplacer, not *builder.BytesWriter

                /home/runner/go/pkg/mod/xorm.io/builder@v0.3.12/builder_limit.go:27 (0xf2bef0)
                	(*Builder).limitWriteTo: ow := w.(*BytesWriter)
                /home/runner/go/pkg/mod/xorm.io/builder@v0.3.12/builder_select.go:148 (0xf2d48b)
                	(*Builder).selectWriteTo: if err := b.limitWriteTo(w); err != nil {
                /home/runner/go/pkg/mod/xorm.io/builder@v0.3.12/builder.go:255 (0xf24896)
                	(*Builder).WriteTo: return b.selectWriteTo(w)
                /home/runner/go/pkg/mod/xorm.io/xorm@v1.3.3-0.20230714073535-9b71cb49cca4/internal/statements/join.go:64 (0x105c3f8)
                	(*Statement).writeJoin: if err := tp.WriteTo(statement.QuoteReplacer(buf)); err != nil {
                /home/runner/go/pkg/mod/xorm.io/xorm@v1.3.3-0.20230714073535-9b71cb49cca4/internal/statements/join.go:30 (0x105bca4)
                	(*Statement).writeJoins: if err := statement.writeJoin(w, join); err != nil {
                /home/runner/go/pkg/mod/xorm.io/xorm@v1.3.3-0.20230714073535-9b71cb49cca4/internal/statements/query.go:193 (0x1060d9c)
                	(*Statement).writeFrom: return statement.writeJoins(w)
                /home/runner/go/pkg/mod/xorm.io/xorm@v1.3.3-0.20230714073535-9b71cb49cca4/internal/statements/query.go:361 (0x1063144)
                	(*Statement).writeSelect: if err := statement.writeFrom(buf); err != nil {
                /home/runner/go/pkg/mod/xorm.io/xorm@v1.3.3-0.20230714073535-9b71cb49cca4/internal/statements/query.go:522 (0x1065504)
                	(*Statement).GenFindSQL: if err := statement.writeSelect(buf, statement.genSelectColumnStr(), true); err != nil {
                /home/runner/go/pkg/mod/xorm.io/xorm@v1.3.3-0.20230714073535-9b71cb49cca4/session_find.go:146 (0x10a70d2)
                	(*Session).find: sqlStr, args, err := session.statement.GenFindSQL(autoCond)
                /home/runner/go/pkg/mod/xorm.io/xorm@v1.3.3-0.20230714073535-9b71cb49cca4/session_find.go:33 (0x10a5844)
                	(*Session).Find: return session.find(rowsSlicePtr, condiBean...)
                /home/runner/work/gitea/gitea/models/packages/nuget/search.go:57 (0x33a73a4)
                	SearchVersions: return pvs, total, sess.Find(&pvs)
```

**Comment by lunny** [comment]
> w.(*BytesWriter)

This issue has been resolved in builder. I just created a new tag and please upgrade builder to v0.3.13

File: models/packages/nuget/search.go (hunk 2 above)
Comment (delvh) [reply models/packages/nuget/search.go]:
```suggestion
		Dialect(db.BuilderDialect()). // Needed for the Limit below
```

File: models/packages/nuget/search.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+		"package.owner_id":    opts.OwnerID,
+		"package.type":        packages_model.TypeNuGet,
+	}
+	if opts.Name.Value != "" {
+		if opts.Name.ExactMatch {
+			cond = cond.And(builder.Eq{"package.lower_name": strings.ToLower(opts.Name.Value)})
+		} else {
+			cond = cond.And(builder.Like{"package.lower_name", strings.ToLower(opts.Name.Value)})
+		}
+	}
```
Comment (delvh) [inline models/packages/nuget/search.go]: I've seen this code twice in this file.
What about a common function?

**Review by wxiaoguang — APPROVED** [review]

**Review by lunny — APPROVED** [review]

**Review by delvh — APPROVED** [review]

**Pull request merged**